pkg/ucp/backend/controller/resourcegroups/trackedresourceprocess.go

```diff
@@ -76,8 +81,18 @@ func (c *TrackedResourceProcessController) Run(ctx context.Context, request *ctr
 		return ctrl.Result{}, fmt.Errorf("failed to validate downstream: %w", err)
 	}
 
+	if downstreamURL == nil {
```

Comment:
could it be empty string too?

Reply:
`downstreamURL` is an `*url.URL`. In general in Go, we can't prevent a function from returning invalid data. The language doesn't give you tools for that.

If someone changes this to return an empty string or empty URL then it will cause tests to fail.